Review (#155)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

claims and likewise, anyone with the decryption key receiving an encrypted JWT
can also read all of the claims.

One of the common use cases of a signed JWT is representing a user's identity
created by an issuer. As long as the signed JWT is one-time use, it typically
only contains those claims the user has consented to disclose to a specific

> verifier. However, when a signed JWT is intended to be multi-use, it needs to

Author: Will fix.

> verifier. However, when a signed JWT is intended to be multi-use, it needs to

Reviewer: I love the new language, thank you!

claims and likewise, anyone with the decryption key receiving an encrypted JWT
can also read all of the claims.

One of the common use cases of a signed JWT is representing a user's
identity. As long as the signed JWT is one-time
use, it typically only contains those claims the user has consented to
disclose to a specific Verifier. However, there is an increasing number
of use cases where a signed JWT is created once and then used a number
of times by the user (the "Holder" of the JWT). In such cases, the signed JWT needs
to contain the superset of all claims the user of the
signed JWT might want to disclose to Verifiers at some point. The
ability to selectively disclose a subset of these claims depending on
the Verifier becomes crucial to ensure minimum disclosure and prevent
Verifiers from obtaining claims irrelevant for the transaction at hand.

One example of such a multi-use JWT is a verifiable credential, a
tamper-evident credential with a cryptographically verifiable authorship that
contains claims about a subject. SD-JWTs defined in this document enable such
selective disclosure of claims.

> In an SD-JWT, claim values are hidden, but cryptographically protected

Reviewer: ```suggestion
In an issuer-issued SD-JWT, claim values are hidden, but cryptographically protected
```

> In an SD-JWT, claim values are hidden, but cryptographically protected

Reviewer: should we use 'digests' in the introduction?

against undetected modification. When issuing the SD-JWT to the Holder,
the issuer also sends a JSON object that contains a mapping between
hidden claim values and their cleartext counterparts, the so-called
Disclosures. This JSON object is therefore called the Issuer-Issued
Disclosures (II-Disclosures) object.

The Holder decides which claims to disclose to a Verifier. This
specification defines a format for conveying the selected subset of the
II-Disclosures to the Verifier. This subset is called the

> Holder-Selected Disclosures and is transported in a JWT, the

Reviewer: ```suggestion
Holder-Selected Disclosures (HS-Disclosures) and is transported in a JWT, the
```

HS-Disclosures JWT, for presentation alongside the SD-JWT. The Verifier

> can verify that all disclosed claim values were part of the original,

Author: ```suggestion
can (and has to) verify that all disclosed claim values were part of the original,
```

issuer-signed SD-JWT. The Verifier will not, however, learn any claim

> values not disclosed in HS-Disclosures.

Reviewer: since that is being pointed out as the difference in security posture btw normal JWT and SD-JWT..

> values not disclosed in HS-Disclosures.

Reviewer: `can and has to`..

While JWTs for claims describing natural persons are a common use case,
the mechanisms defined in this document can be used for many other use
cases as well.

This document also describes an optional mechanism for Holder Binding,
or the concept of binding an SD-JWT to key material controlled by the
Holder.

> This specification aims to be easy to imeplement and to leverage

Reviewer: ```suggestion
This specification aims to be easy to implement and to leverage
```

has consented to disclose to the Verifier. This enables the Verifier to
verify the claims received from the Holder by computing the digests of
the claim values, salts, and potentially cleartext claim names revealed
in the HS-Disclosures JWT using the digest derivation algorithm
specified in SD-JWT and comparing them to the digests included in
SD-JWT.

The Disclosures are contained in the `sd_hs_disclosures` object. The
structure of the `sd_hs_disclosures` object in the HS-Disclosures JWT is
the same as the structure of the `sd_ii_disclosures` object in the
II-Disclosures Object, but any claims the Holder wishes not to disclose
are omitted.

The HS-Disclosures JWT MAY contain further claims, for example, to ensure a binding
to a concrete transaction (in the example below, the `nonce` and `aud` claims).

When the Holder sends the HS-Disclosures JWT to the Verifier, the HS-Disclosures JWT MUST be a JWS
represented as the JWS Compact Serialization as described in
Section 7.1 of [@!RFC7515].

If Holder Binding is desired, the HS-Disclosures JWT is signed by the Holder. If no
Holder Binding is to be used, the `none` algorithm is used, i.e., the document

> is not signed. As described in (#verifier-verification), a verified MUST NOT accept an HS-Disclosures JWT with `alg=none` or without a signature unless Holder Binding is not required. See also (#holder_binding_security).

Reviewer: ```suggestion
is not signed. 

Whether to check the signature of the HS-Disclosures JWT is up to the Verifier's policy,
based on the set of trust requirements such as trust frameworks it belongs to. 
As described in (#verifier-verification), the Verifier MUST NOT accept HS-Disclosures 
JWTs using "none" algorithm, when the Verifier's policy requires a signed 
HS-Disclosures JWT. See also (#holder_binding_security).
```

Reviewer: Ideally, I would like this paragraph to be in a more easily finable section....

H-ZtWw
```

## Combined Format for Presentation

> The SD-JWT and the HS-Disclosures JWT can be combined into one document using period character `.` as a separator. This means that the resulting string consists of six dot-separated parts as follows, where the last part may be empty:

Reviewer: ```suggestion
The SD-JWT and the HS-Disclosures JWT can be combined into one document
using period character `.` as a separator. This means that the resulting string consists of
six dot-separated parts as described below.

 The last part (HSD Signature) may be empty when Holder Binding is not used and 
 HS-Disclosures JWT is not signed.
```

# Verification and Processing

## Verification by the Holder when Receiving SD-JWT and Issuer-Issued Disclosures Object {#holder-verification}

The Holder SHOULD verify the binding between SD-JWT and II-Disclosures Object by performing the following steps:
 1. Check that all the claims in the II-Disclosures Object are present in the SD-JWT and that there are no claims in the SD-JWT that are not in the II-Disclosures Object
 2. Check that the digests of the claims in the II-Disclosures Object match those in the SD-JWT

## Verification by the Verifier when Receiving SD-JWT and Holder-Selected Disclosures JWT {#verifier-verification}

Verifiers MUST follow [@RFC8725] for checking the SD-JWT and, if signed, the
HS-Disclosures JWT.

Verifiers MUST go through (at least) the following steps before
trusting/using any of the contents of an SD-JWT:

>  1. Determine if Holder Binding is to be checked for the use case at
>     hand. The decision MUST NOT be based on whether the HS-Disclosures

Reviewer: ```suggestion
 1. Determine if Holder Binding is to be checked according to the Verifier's policy
    for the use case at hand. This decision MUST NOT be based on whether 
    the HS-Disclosures
```

revealed fundamental weaknesses and they MUST NOT be used.

## Holder Binding {#holder_binding_security}
Verifiers MUST decide whether Holder Binding is required for a
particular use case or not before verifying a credential. This decision

> can be informed by business requirements, the use case, the type of

Reviewer: ```suggestion
can be informed by various factors including, but not limited to the following:
business requirements, the use case, the type of
```

binding between a Holder and its credential that is required for a use
case, the sensitivity of the use case, the expected properties of a
credential, the type and contents of other credentials expected to be
presented at the same time, etc.

> This can be showcased on two use cases based on a mobile driver's license:

Reviewer: ```suggestion
This can be showcased based on a mobile driver's license use case for SD-JWT:
```

> **Use case A:** For the verification of the driver's license at a

Reviewer: ```suggestion
**Scenario A:** For the verification of the driver's license at a
```

> **Use case A:** For the verification of the driver's license at a
> traffic stop, Holder Binding may be necessary to ensure that the person

Reviewer: ```suggestion
**Scenario A:** For the verification of the driver's license when
stopped by a police officer for exceeding a speed limit, Holder Binding may be necessary to ensure that the person
```

   * `sd_disclosure` in II-Disclosures renamed to `sd_ii_disclosures`
   * `sd_disclosure` in HS-Disclosures renamed to `sd_hs_disclosures`
   * clarified relationship between `sd_hs_disclosure` and SD-JWT
   * clarified combined formats for issuance and presentation

>    * improved description of holder binding security considerations

Reviewer: ```suggestion
```

   * `sd_disclosure` in II-Disclosures renamed to `sd_ii_disclosures`
   * `sd_disclosure` in HS-Disclosures renamed to `sd_hs_disclosures`
   * clarified relationship between `sd_hs_disclosure` and SD-JWT
   * clarified combined formats for issuance and presentation
   * clarified security requirements for blinded claim names

>    * improved description of Holder Binding security considerations

Reviewer: ```suggestion
   * improved description of Holder Binding security considerations - expecially around the usage of "alg=none".
```

   *  Editorial fixes
   *  Added `hash_alg` claim
   *  Renamed `_sd` to `sd_digests` and `sd_release`

>    *  Added descriptions on Holder Binding - more work to do

Reviewer: ```suggestion
   *  Added descriptions on the security of Holder Binding -  expecially around the usage of "alg=none".
```

   *  Clarify that signing the SD-JWT is mandatory

   -00

   *  Renamed to SD-JWT (focus on JWT instead of JWS since signature is optional)

>    *  Make Holder Binding optional
>    *  Rename proof to release, since when there is no signature, the term "proof" can be misleading

Reviewer: ```suggestion
   *  Made Holder Binding optional
   *  Renamed proof to release, since when there is no signature, the term "proof" can be misleading
```
